src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
             return false;
         }
 
+        // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
+        // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
+        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        private SumType<CompletionItem[], CompletionList>? RemoveDesignTimeItems(
+            SumType<CompletionItem[], CompletionList> completionResult,
+            LSPDocumentSnapshot documentSnapshot,
+            TextExtent? wordExtent,
+            LanguageServerKind serverKind)
+        {
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                return completionResult;
+            }
+
+            var result = completionResult.Match<SumType<CompletionItem[], CompletionList>?>(
+                items => RemoveDesignTimeItems(items, documentSnapshot, wordExtent),
+                list => RemoveDesignTimeItems(list.Items, documentSnapshot, wordExtent));
+
+            return result;
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
+            {
+                var designTimeHelpersCompletionItems = GenerateCompletionItems(DesignTimeHelpers);
+                if (wordExtent.HasValue)
+                {
+                    var wordCharacterStartIndex = wordExtent.Value.Span.Start.Position;
+                    if (wordCharacterStartIndex + 1 < documentSnapshot.Snapshot.Length)
+                    {
+                        var snapshot = documentSnapshot.Snapshot;
+
+                        // If the current identifier starts with "__", only trim out common design time helpers from the list.
+                        if (snapshot[wordCharacterStartIndex] == '_' && snapshot[wordCharacterStartIndex + 1] == '_')
+                        {
+                            return items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+                        }
+                    }
+                }
+
+                // In all other cases, trim out both common design time helpers and all completion items starting with "__".
+                var updatedItems = items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+                updatedItems = updatedItems.Where(item => item.InsertText != null && !item.InsertText.StartsWith("__", StringComparison.Ordinal)).ToArray();
+
+                return updatedItems;
+            }
+        }
+
+        private static IReadOnlyCollection<CompletionItem> GenerateCompletionItems(IReadOnlyCollection<string> completionItems)
+            => completionItems.Select(k => new CompletionItem
+            {
+                Label = k,
+                InsertText = k,
```

Comment:
No need to provide InsertText, filterText or SortText, they default to Label according to the spec. Also don't need to provide the InserTextFormat